### PR TITLE
WIP evolving the IR

### PR DESCRIPTION
**Pull request by JeffBezanson**
This represents local variables as `Slot(i)` instead of symbols, making parts of the compiler simpler and hopefully faster. Also helps the interpreter.

~~Status: builds a working system image, but only with inlining disabled. Inlining pass seems not to fully work yet.~~

cc @vtjnash @carnaval 

Remaining steps:
- [x] Get everything working with numbered slots
- [x] Convert varinfo from array-of-arrays to 3 arrays
- [ ] Switch from Expr to compact types for assignment, return, gotoifnot, call{n}
- [x] Don't store mangled and compiler-generated symbol names in varinfo
- [ ] A more efficient way to annotate uses of slots with types. There are too many `Slot` objects, and the vast majority ought to be shared like GenSyms are.

### CODE REVIEW

File: src/interpreter.c [hunk 1]
```diff
@@ -14,34 +14,27 @@
...
-    return eval(e, NULL, 0, 0);
+    return eval(e, NULL, NULL);
 }
 
-JL_DLLEXPORT jl_value_t *jl_interpret_toplevel_expr_in(jl_module_t *m,
-                                                       jl_value_t *e,
-                                                       jl_svec_t *local_syms,
-                                                       jl_svec_t *local_vals)
+JL_DLLEXPORT jl_value_t *jl_interpret_toplevel_expr_in(jl_module_t *m, jl_value_t *e,
+                                                       jl_lambda_info_t *lam)
```
Comment (vtjnash) [inline src/interpreter.c]: this seems to assume that `local_vals == lam->sparam_vals`, but that is false for the LambdaInfo reached at `LambdaInfo->unspecialized` and `LambdaInfo`)

Comment (JeffBezanson) [inline src/interpreter.c]: Ok, will add another argument for this.

**Comment by iamed2** [comment]
> Switch from Expr to compact types for assignment, return, gotoifnot, enter, leave, static_parameter

I'm excited for the macro dispatch opportunities this will provide!

**Comment by carnaval** [comment]
this stage happens after the macro expansion

**Comment by carnaval** [comment]
@JeffBezanson might that be a good time to also have varinfo be some kind of type with proper field names so that part of all the magic indices everywhere can go away ? not hugely important but might be a nicety along the way

**Comment by JeffBezanson** [comment]
@carnaval Yes, that's on the list, but I was thinking of using 3 arrays indexed by slot number. That way the flags can be a UInt8 array.

**Comment by JeffBezanson** [comment]
Ok, things are passing, except for one test where the interpreter doesn't yet know the names of local variables, which is expected. The system image build is about 50 seconds faster than master, so this is promising.

**Comment by timholy** [comment]
Wow, 50 seconds. Just fantastic work you folks are doing to improve the core infrastructure.

**Pull request closed without merging**